website/files/models/base.py

```diff
@@ -0,0 +1,588 @@
+
+import os
+import bson
+import logging
+import requests
+
+from modularodm import fields, Q
+from modularodm.exceptions import NoResultsFound
+from dateutil.parser import parse as parse_date
+
+from framework.guid.model import Guid
+from framework.mongo import StoredObject
+from framework.mongo.utils import unique_on
+from framework.analytics import get_basic_counters
+
+from website import util
+from website.files import utils
+from website.files import exceptions
+
+
+__all__ = (
+    'File',
+    'Folder',
+    'FileNode',
+    'FileVersion',
+    'StoredFileNode',
+    'TrashedFileNode',
+)
+
+
+PROVIDER_MAP = {}
+logger = logging.getLogger(__name__)
+
+
+class TrashedFileNode(StoredObject):
+    """The graveyard for all deleted FileNodes"""
+    _id = fields.StringField(primary=True)
+
+    versions = fields.ForeignField('FileVersion', list=True)
+
+    node = fields.ForeignField('node', required=True)
+    parent = fields.AbstractForeignField(default=None)
+
+    is_file = fields.BooleanField(default=True)
+    provider = fields.StringField(required=True)
+
+    name = fields.StringField(required=True)
+    path = fields.StringField(required=True)
+    materialized_path = fields.StringField(required=True)
+
+
+@unique_on(['node', 'name', 'parent', 'is_file', 'provider', 'materialized_path'])
+class StoredFileNode(StoredObject):
+    """The storage backend for FileNode objects.
+    This class should generally not be used or created manually as FileNode
+    contains all the helpers required.
+    A FileNode wraps a StoredFileNode to provider usable abstraction layer
+    """
+    _id = fields.StringField(primary=True, default=lambda: str(bson.ObjectId()))
+
+    versions = fields.ForeignField('FileVersion', list=True)
+
+    node = fields.ForeignField('Node', required=True)
+    parent = fields.ForeignField('StoredFileNode', default=None)
+
+    is_file = fields.BooleanField(default=True)
+    provider = fields.StringField(required=True)
+
+    name = fields.StringField(required=True)
+    path = fields.StringField(required=True)
+    materialized_path = fields.StringField(required=True)
+
+    @property
+    def deep_url(self):
+        return self.wrapped().deep_url
+
+    def wrapped(self):
+        """Wrap self in a FileNode subclass
+        """
+        return PROVIDER_MAP[self.provider][int(self.is_file)](self)
+
+    def get_guid(self, create=False):
+        """Attempt to find a Guid that points to this object.
+        One will be created if requested.
+        :rtype: Guid
+        """
+        try:
+            return Guid.find_one(Q('referent', 'eq', self))
+        except NoResultsFound:
+            if not create:
+                return None
+        return Guid.generate(self)
+
+
+class FileNodeMeta(type):
+    """Keeps track of subclasses of the ``FileNode`` object
+    Inserts all into the PROVIDER_MAP following the pattern:
+    {
+        provider: [ProviderFolder, ProviderFile, ProviderFileNode]
```

Comment:
Note: I don't feel strongly that it should be changed; just trying to understand the rationale.